File: solutions/csharp/bowling/1/Bowling.cs
public class BowlingGame
{
    private readonly List<int> _rolls = [];
    private readonly List<FrameScore> _frames = [];
    private GameState _state = new FirstRoll(Frame: 1);

    public void Roll(int pins)
    {
        if (pins is < 0 or > 10)
            throw new ArgumentException();

        var (newState, score) = _state.Next(pins, _rolls.Count);

        _state = newState;
        _rolls.Add(pins);

        if (score is not null)
            _frames.Add(score);
    }

    public int? Score()
    {
        if (_state is not Complete)
            throw new ArgumentException();

        return _frames
            .Sum(frame => frame.Score(_rolls));
    }

    private static void EnsureTwoRollsDontExceedTen(int first, int second)
    {
        if (first + second > 10)
            throw new ArgumentException();
    }

    private abstract record GameState
    {
        public abstract (GameState State, FrameScore? Score) Next(int pins, int rollIndex);
    }

    private sealed record FirstRoll(int Frame) : GameState
    {
        public override (GameState State, FrameScore? Score) Next(int pins, int rollIndex) =>
            pins switch
            {
                10 when Frame == 10 => (new TenthStrikeBonusFirst(), new StrikeScore(rollIndex)),
                10 => (new FirstRoll(Frame + 1), new StrikeScore(rollIndex)),
                _ when Frame == 10 => (new TenthSecondRoll(pins, rollIndex), null),
                _ => (new SecondRoll(Frame, pins, rollIndex), null)
            };
    }

    private sealed record SecondRoll(int Frame, int First, int FrameIndex) : GameState
    {
        public override (GameState State, FrameScore? Score) Next(int pins, int rollIndex)
        {
            EnsureTwoRollsDontExceedTen(First, pins);

            FrameScore score = First + pins == 10
                ? new SpareScore(FrameIndex)
                : new OpenFrameScore(FrameIndex);

            return (new FirstRoll(Frame + 1), score);
        }
    }

    private sealed record TenthSecondRoll(int First, int FrameIndex) : GameState
    {
        public override (GameState State, FrameScore? Score) Next(int pins, int rollIndex)
        {
            EnsureTwoRollsDontExceedTen(First, pins);

            return First + pins == 10
                ? (new TenthSpareBonus(FrameIndex), null)
                : (new Complete(), new OpenFrameScore(FrameIndex));
        }
    }

    private sealed record TenthSpareBonus(int FrameIndex) : GameState
    {
        public override (GameState State, FrameScore? Score) Next(int pins, int rollIndex) =>
            (new Complete(), new SpareScore(FrameIndex));
    }

    private sealed record TenthStrikeBonusFirst : GameState
    {
        public override (GameState State, FrameScore? Score) Next(int pins, int rollIndex) =>
            (new TenthStrikeBonusSecond(pins), null);
    }

    private sealed record TenthStrikeBonusSecond(int FirstBonus) : GameState
    {
        public override (GameState State, FrameScore? Score) Next(int pins, int rollIndex)
        {
            if (FirstBonus != 10 && FirstBonus + pins > 10)
                throw new ArgumentException();

            return (new Complete(), null);
        }
    }

    private sealed record Complete : GameState
    {
        public override (GameState State, FrameScore? Score) Next(int pins, int rollIndex) =>
            throw new ArgumentException();
    }

    private abstract record FrameScore(int Index)
    {
        public abstract int Score(IReadOnlyList<int> rolls);
    }

    private abstract record ThreeRollScore(int Index) : FrameScore(Index)
    {
        public override int Score(IReadOnlyList<int> rolls) => rolls[Index] + rolls[Index + 1] + rolls[Index + 2];
    }

    private sealed record StrikeScore(int Index) : ThreeRollScore(Index);

    private sealed record SpareScore(int Index) : ThreeRollScore(Index);

    private sealed record OpenFrameScore(int Index) : FrameScore(Index)
    {
        public override int Score(IReadOnlyList<int> rolls) => rolls[Index] + rolls[Index + 1];
    }
}

File: solutions/python/bowling/1/bowling.py
"""Score bowling games with explicit roll states and frame scoring windows."""

from enum import Enum, auto


class _State(Enum):
    FIRST_ROLL = auto()
    SECOND_ROLL = auto()
    SPARE_BONUS = auto()
    STRIKE_BONUS_FIRST = auto()
    STRIKE_BONUS_SECOND = auto()
    COMPLETE = auto()


class BowlingGame:
    """Track legal rolls and score a completed ten-frame game."""

    def __init__(self):
        self._rolls = []
        self._frames = []
        self._state = _State.FIRST_ROLL
        self._first = 0

    def roll(self, pins):
        """Record a roll, rejecting invalid pins or rolls after completion."""
        self._validate_roll(pins)
        self._advance(pins)
        self._rolls.append(pins)

    def score(self):
        """Return the final score once all frames and fill balls are complete."""
        if self._state != _State.COMPLETE:
            raise ValueError("cannot score an incomplete game")
        return sum(sum(self._rolls[frame]) for frame in self._frames)

    def _validate_roll(self, pins):
        self._validate_pin_range(pins)
        self._validate_game_in_progress()
        self._validate_standing_pins(pins)

    @staticmethod
    def _validate_pin_range(pins):
        if not 0 <= pins <= 10:
            raise ValueError("pins must be between 0 and 10")

    def _validate_game_in_progress(self):
        if self._state == _State.COMPLETE:
            raise ValueError("cannot roll after the game is complete")

    def _validate_standing_pins(self, pins):
        if (
            self._state == _State.SECOND_ROLL
            or (self._state == _State.STRIKE_BONUS_SECOND and self._first != 10)
        ) and self._first + pins > 10:
            raise ValueError("roll exceeds the number of standing pins")

    def _advance(self, pins):
        match self._state:
            case _State.FIRST_ROLL:
                self._start_frame(pins)
            case _State.SECOND_ROLL:
                self._finish_frame(pins)
            case _State.STRIKE_BONUS_FIRST:
                self._first = pins
                self._state = _State.STRIKE_BONUS_SECOND
            case _State.SPARE_BONUS | _State.STRIKE_BONUS_SECOND:
                self._state = _State.COMPLETE

    def _start_frame(self, pins):
        if pins == 10:
            index = len(self._rolls)
            self._frames.append(slice(index, index + 3))
            if len(self._frames) == 10:
                self._state = _State.STRIKE_BONUS_FIRST
        else:
            self._first = pins
            self._state = _State.SECOND_ROLL

    def _finish_frame(self, pins):
        is_spare = self._first + pins == 10
        index = len(self._rolls) - 1
        self._frames.append(slice(index, index + (3 if is_spare else 2)))
        if len(self._frames) == 10:
            self._state = _State.SPARE_BONUS if is_spare else _State.COMPLETE
        else:
            self._state = _State.FIRST_ROLL
